src/io/mouse.js

```diff
@@ -1,20 +1,47 @@
 var MathUtil = require('../util/math-util');
 
-function Mouse () {
+function Mouse (runtime) {
     this._x = 0;
     this._y = 0;
     this._isDown = false;
+    /**
+     * Reference to the owning Runtime.
+     * Can be used, for example, to activate hats.
+     * @type{!Runtime}
+     */
+    this.runtime = runtime;
 }
 
 Mouse.prototype.postData = function(data) {
     if (data.x) {
-        this._x = data.x;
+        this._x = data.x - data.canvasWidth / 2;
     }
     if (data.y) {
-        this._y = data.y;
+        this._y = data.y - data.canvasHeight / 2;
     }
     if (typeof data.isDown !== 'undefined') {
         this._isDown = data.isDown;
+        if (this._isDown) {
+            this._activateClickHats(data.x, data.y);
+        }
+    }
+};
+
+Mouse.prototype._activateClickHats = function (x, y) {
```

Comment:
I think I'm starting to come around to what I suspect is your point of view. I think I had a different "meaning" for I/O device than you had, and I got confused when you followed your mental model instead of mine ;)

I had been thinking of the I/O devices as a low-level data source, and that concept gets pretty weak when you introduce a central message dispatcher. That's where my "I feel like this code should be at a higher level" comment came from, as well as my thoughts about translating messages into events.

However, if you think of an I/O device class as a place to handle data related to a particular device, then it makes sense to have this hat block code in there. The message passing is there sort of as a consequence of the fact that we don't currently have another way to handle that. If we add a central message dispatcher or even stop using a worker (#165) for the VM, then this class is still valuable as a place to put mouse- (or pointer-) specific work.

I think I still like the idea of centralizing message-to-event conversion if it turns out to make sense, but I now think that would just clarify the purpose of `src/io/*.js` instead of replacing them.

Regarding the cases you mentioned:
- Video sensing: I suspect we'll want to detect motion on the main thread in order to avoid passing the whole video stream as messages. Once the data gets to the VM side, it could be caught by a "video I/O" device which then dispatches Scratch events if appropriate.
- Microphone sensing: Similar, as you suggested.
- Variables: `src/io/cloud.js`? It feels weird initially but really it's doing the same sort of thing. It makes me feel like "I/O device" might not be the right term for these.
- Audio & stage rendering: perhaps the "remote control" (`RenderWebGLWorker.js` for example) should be an I/O device, or should be wrapped by a thin I/O device? I could see that making sense.

Sorry for the disconnect. Does this seem like we're closer to being on the same page?